File: src/main/java/core/scm/MapEntry.java
package core.scm;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.writer.Writer;

import java.util.Map;

public class MapEntry extends AFn implements Map.Entry {

  private final Object key;
  private final Object val;

  public MapEntry(Object key, Object val) {
    super(new FnArgsBuilder().min(1).max(1).mandatory(new Class[]{Type.ExactNonNegativeInteger.class}).build());
    this.key = key;
    this.val = val;
  }

  @Override
  public Object apply1(Object arg) {
    int index = ((Number)arg).intValue();
    if (index > 2) {
      throw new IndexOutOfBoundsException(getName() + ": value out of range: " + index);
    }
    return index == 0 ? key : val;
  }

  @Override
  public Object getKey() {
    return key;
  }

  @Override
  public Object getValue() {
    return val;
  }

  @Override
  public Object setValue(Object value) {
    throw new UnsupportedOperationException();
  }

  @Override
  public String getName() {
    return "map entry";
  }

  @Override
  public String toString() {
    return "[" + Writer.write(key) + " " + Writer.write(val) + "]";
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    MapEntry that = (MapEntry) o;
    if (key != null ? !key.equals(that.key) : that.key != null) return false;
    return val != null ? val.equals(that.val) : that.val == null;
  }

  @Override
  public int hashCode() {
    int result = key != null ? key.hashCode() : 0;
    result = 31 * result + (val != null ? val.hashCode() : 0);
    return result;
  }
}

File: src/main/java/core/scm/MapEntry.kt
package core.scm

import core.writer.Writer

// TODO implement proper interfaces, refactor
class MapEntry(key: Any?, value: Any?) : MutableVector(key, value) {

    constructor(entry: Map.Entry<Any?, Any?>) : this(entry.key, entry.value)

    val key: Any?
        get() = get(0)

    val value: Any?
        get() = get(1)

    override val name: String
        get() = "map entry"

    override fun toString(): String {
        return "[" + Writer.write(key) + " " + Writer.write(value) + "]"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as MapEntry?
        if (if (key != null) key != that!!.key else that!!.key != null) return false
        return if (value != null) value == that.value else that.value == null
    }

    override fun hashCode(): Int {
        var result = if (key != null) key!!.hashCode() else 0
        result = 31 * result + if (value != null) value!!.hashCode() else 0
        return result
    }
}